server/client.go

```diff
@@ -1428,13 +1454,28 @@ func (c *client) canSubscribe(subject []byte) bool {
 
 	// Check allow list. If no allow list that means all are allowed. Deny can overrule.
 	if c.perms.sub.allow != nil {
-		r := c.perms.sub.allow.Match(string(subject))
+		r := c.perms.sub.allow.Match(subject)
 		allowed = len(r.psubs) != 0
 	}
 	// If we have a deny list and we think we are allowed, check that as well.
 	if allowed && c.perms.sub.deny != nil {
-		r := c.perms.sub.deny.Match(string(subject))
+		r := c.perms.sub.deny.Match(subject)
 		allowed = len(r.psubs) == 0
+
+		// We use the actual subscription to signal us to spin up the deny mperms
+		// and cache. We check if the subject is a wildcard that contains any of
+		// the deny clauses.
+		// FIXME(dlc) - We could be smarter and track when these go away and remove.
+		if c.mperms == nil && subjectHasWildcard(subject) {
```

Comment:
Should remove `c.mperms == nil`. We need to think of reload. On reload, we will call processSubsOnConfigReload() for each client and that calls canSubscribe(). If there was a deny before but not anymore, this would need to be cleared, etc..
But even then, it brings the question: if a client has already a mperms, then any other subscription for that client would not be making this check? Why? Say that you have deny clauses on "foo.bar" and "foo.baz", and for the first time create a subscription on "*.bar", then only "foo.bar" would be added to the mperms. But if later I create a subscription on ">" or "*.baz", etc.. then I should also have "foo.baz" in mperms.

Reply:
I agree need to handle reload, I will think through that. My first thought is just clear mperms on processSubsOnConfigReload() which will then reprocess all the subs that we have as if they are being recreated. 

For not checking, see comment above. Once we get triggered we load all deny filters at once, not one at a time.

Reply:
Yes, clearing prior to call canSubscribe() would repopulate it I guess.

Reply:
Will make that change..

Reply:
Right now we don't hold the lock for canSubscribe() in that function because, as the comment say, the only way perms could change is during reload, which in this case we are called from.

But now, since we may create mperms in canSubscribe(), I think that you must lock around canSubscribe() in  processSubsOnConfigReload (), and so you could set it to nil before actual call to canSubscribe().

Reply:
Safe to hold the lock in your opinion there?

Reply:
What I meant is in processSubsOnConfigReload(), you could do something like:
```
for _, sub := range subs {
    if checkPerms {
        c.mu.Lock()
        c.mperms = nil
        canSub := c.canSubscribe(sub.subject)
        c.mu.Unlock()
        if !canSub {
            removed = append(removed, sub)
	    c.unsubscribe(acc, sub, true)
        }
    }
    if checkAcc {
        c.mu.Lock()
...
```

Reply:
Can't do that, need to clear it once before looping through subs.

Reply:
Oh yes, it is per client, while this loop is for all subs for that client.

Reply:
I created another solution that I think works well.